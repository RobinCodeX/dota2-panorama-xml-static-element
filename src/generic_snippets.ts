import { DOMParser, XMLSerializer } from '@xmldom/xmldom';
import { readFile } from 'fs/promises';

async function LoadSnippetsFile(file: string) {
    const source = await readFile(file, 'utf8');
    const doc = new DOMParser().parseFromString(source, 'text/xml');
    return doc.getElementsByTagName('snippets').item(0);
}

/**
 * Return all snippets
 */
export async function MakeGenericSnippets(files: string[]): Promise<Node[]> {
    if (files.length <= 0) {
        return [];
    }
    const result: Node[] = [];
    const snippetsList = await Promise.all(files.map(LoadSnippetsFile));
    for (const snippets of snippetsList) {
        if (!snippets) {
            continue;
        }
        for (let i = 0; i < snippets.childNodes.length; i++) {
            const element = snippets.childNodes.item(i) as Element;
            if (
                element &&
                element.nodeType === element.ELEMENT_NODE &&
                element.tagName === 'snippet'
            ) {
                result.push(element.cloneNode(true));
            }
        }
    }
    return result;
}
