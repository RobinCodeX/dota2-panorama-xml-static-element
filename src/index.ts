import { DOMParser, XMLSerializer } from 'xmldom';
import { promises } from 'fs';
import formatter from 'xml-formatter';

interface RenderContext {
    readonly RootFile: string;
    readonly Root: Document;
    readonly FileDoc: Record<string, Document>;
}

const emptyStringMathcer = /^\s+$/;

function getFirstElement(root: Node): Element | undefined {
    let child = root.firstChild;
    while (child) {
        if (child.nodeType === child.ELEMENT_NODE) {
            return child as Element;
        }
        child = child.nextSibling;
    }
}

function queryElementsByAttribute(attrName: string, root: Node): Element[] {
    const list: Element[] = [];
    let child = root.firstChild;
    while (child) {
        if (child.nodeType === child.ELEMENT_NODE) {
            if (!!(child as Element).getAttribute(attrName)) {
                list.push(child as Element);
            } else if (child.hasChildNodes()) {
                queryElementsByAttribute(attrName, child);
            }
        }
        child = child.nextSibling;
    }
    return list;
}

function queryElementsByID(id: string, root: Node): Element | undefined {
    let child = root.firstChild;
    while (child) {
        if (child.nodeType === child.ELEMENT_NODE) {
            if ((child as Element).getAttribute('id') === id) {
                return child as Element;
            } else if (child.hasChildNodes()) {
                const nextChild = queryElementsByID(id, child);
                if (nextChild) {
                    return nextChild;
                }
            }
        }
        child = child.nextSibling;
    }
}

function getAllAttributes(root: Element): Record<string, string> {
    const result: Record<string, string> = {};
    for (let i = 0; i < root.attributes.length; i++) {
        const attr = root.attributes.item(i)!;
        result[attr.name] = attr.value;
    }
    return result;
}

function replaceVariables(root: Element, key: string, value: string) {
    if (root.nodeType === root.ELEMENT_NODE) {
        for (let i = 0; i < root.attributes.length; i++) {
            const attr = root.attributes.item(i)!;
            attr.value = attr.value.replace(`@${key}`, value);
        }
    }

    if (root.hasChildNodes()) {
        let child = root.firstChild as Element;
        while (child) {
            if (child.nodeType === child.ELEMENT_NODE) {
                replaceVariables(child, key, value);
            }
            child = child.nextSibling as Element;
        }
    }
}

function ClearComments(root: Node) {
    for (let i = root.childNodes.length - 1; i >= 0; i--) {
        const child = root.childNodes.item(i);
        if (child.nodeType === child.COMMENT_NODE) {
            child.parentNode?.removeChild(child);
        } else if (child.hasChildNodes()) {
            ClearComments(child);
        }
    }
}

function ReplaceTemplates(templates: Element[], root: Node) {
    for (let i = root.childNodes.length - 1; i >= 0; i--) {
        let child = root.childNodes.item(i) as Element;
        if (child.nodeType === child.ELEMENT_NODE) {
            const template = templates.find((v) => v.getAttribute('name') === child.nodeName);
            if (template) {
                const firstElement = getFirstElement(template.cloneNode(true));
                if (!firstElement) {
                    throw Error(`not first child in template ${child.nodeName}`);
                }

                // Copy children
                const bindElements = queryElementsByAttribute('bind', firstElement);
                for (const elem of bindElements) {
                    const bindTag = elem.getAttribute('bind')!;
                    elem.removeAttribute('bind');
                    const target = child.getElementsByTagName(bindTag).item(0);
                    if (target) {
                        for (let i = 0; i < target.childNodes.length; i++) {
                            const element = target.childNodes.item(i);
                            elem.appendChild(element.cloneNode(true));
                        }
                    }
                }

                for (const [k, v] of Object.entries(getAllAttributes(child))) {
                    replaceVariables(firstElement, k, v);
                }

                // Replace element
                child.parentNode?.replaceChild(firstElement, child);
                child = firstElement;
            }
            if (child.hasChildNodes()) {
                ReplaceTemplates(templates, child);
            }
        }
    }
}

function StartRender(ctx: RenderContext) {
    ClearComments(ctx.Root);

    // Find templates
    const templateElements = ctx.Root.getElementsByTagName('Template');
    const templates: Element[] = [];
    for (let i = templateElements.length - 1; i >= 0; i--) {
        const element = templateElements.item(i)!;
        element.parentNode?.removeChild(element);
        templates.push(element);
    }

    ReplaceTemplates(templates, ctx.Root);
}

export interface RenderPanoramaXMLOptions {
    indentation: string;
}

export async function RenderPanoramaXML(
    xmlFilePath: string,
    options?: RenderPanoramaXMLOptions
): Promise<string> {
    if (!options) {
        options = { indentation: '  ' };
    }

    const body = await promises.readFile(xmlFilePath, { encoding: 'utf8' });
    const doc = new DOMParser().parseFromString(body, 'text/xml');
    const ctx: RenderContext = {
        RootFile: xmlFilePath,
        Root: doc,
        FileDoc: { [xmlFilePath]: doc },
    };
    StartRender(ctx);

    return formatter(new XMLSerializer().serializeToString(ctx.Root), {
        indentation: options.indentation || '  ',
        lineSeparator: '\n',
    });
}

export default RenderPanoramaXML;
