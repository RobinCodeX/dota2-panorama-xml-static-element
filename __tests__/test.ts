import { writeFile } from 'fs/promises';
import path = require('path');
import { PreloadTemplates, RenderPanoramaXML } from '../src/index';

describe('Test xml render', () => {
    test('test.xml', async () => {
        const result = await RenderPanoramaXML(path.join(__dirname, 'test.xml'), {
            indentation: '    ',
        });
        expect(result).toMatchSnapshot();
    });

    test('test.xml no options', async () => {
        const result = await RenderPanoramaXML(path.join(__dirname, 'test.xml'));
        expect(result).toMatchSnapshot();
    });

    test('test.xml templateRoots option', async () => {
        const result = await RenderPanoramaXML(path.join(__dirname, 'test.xml'), {
            templateRoots: [path.join(__dirname, 'components')],
        });
        expect(result).toMatchSnapshot();
    });

    test('test.xml template not root element', async () => {
        const result = await RenderPanoramaXML(path.join(__dirname, 'test.xml'), {
            templateRoots: [path.join(__dirname, 'no_root_component')],
        });
        expect(result);
    });

    test('preload components', async () => {
        const elements = await PreloadTemplates([path.join(__dirname, 'components')]);
        expect(4).toBe(elements.length);
        expect('Template').toBe(elements[0].tagName);
        expect('Template').toBe(elements[1].tagName);
        expect('DefaultButton').toBe(elements[0].getAttribute('name'));
        expect('HeroButton').toBe(elements[1].getAttribute('name'));
        expect('Button02').toBe(elements[2].getAttribute('name'));
        expect('Button01').toBe(elements[3].getAttribute('name'));

        await expect(
            PreloadTemplates([
                path.join(__dirname, 'components'),
                path.join(__dirname, 'some_componenets'),
            ])
        ).rejects.toThrow(
            `${path
                .join(__dirname, 'some_componenets/DefaultButton.xml')
                .replace(/\\/g, '/')}: Duplicate template name 'DefaultButton'`
        );
    });

    test('generic snippets', async () => {
        const result = await RenderPanoramaXML(path.join(__dirname, 'test_snippets.xml'), {
            snippetsFiles: [
                path.join(__dirname, 'snippets/Ability.xml'),
                path.join(__dirname, 'snippets/Item.xml'),
            ],
        });
        expect(result).toMatchSnapshot();
    });

    test('generic snippets - no snippets', async () => {
        const result = await RenderPanoramaXML(path.join(__dirname, 'test_snippets2.xml'), {
            snippetsFiles: [
                path.join(__dirname, 'snippets/Ability.xml'),
                path.join(__dirname, 'snippets/Item.xml'),
                path.join(__dirname, 'snippets/NoSnippets.xml'),
            ],
        });
        expect(result).toMatchSnapshot();
    });
});
