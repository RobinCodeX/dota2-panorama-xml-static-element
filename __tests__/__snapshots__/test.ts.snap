// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test xml render generic snippets - no snippets 1`] = `
"<root>
  <styles>
    <include src=\\"file://{resources}/styles/test.css\\"/>
  </styles>
  <scripts>
    <include src=\\"file://{resources}/scripts/test.js\\"/>
  </scripts>
  <snippets>
    <snippet name=\\"Ability\\">
      <Panel class=\\"Ability\\">
        <DOTAAbilityImage id=\\"AbilityImage\\"/>
      </Panel>
    </snippet>
    <snippet name=\\"AbilityLarge\\">
      <Panel class=\\"AbilityLarge\\">
        <DOTAAbilityImage id=\\"AbilityImage\\"/>
      </Panel>
    </snippet>
    <snippet name=\\"Item\\">
      <Panel class=\\"Item\\">
        <DOTAItemImage id=\\"ItemImage\\"/>
      </Panel>
    </snippet>
  </snippets>
  <Panel class=\\"test_root\\">
    <Panel/>
    <Panel/>
  </Panel>
</root>"
`;

exports[`Test xml render generic snippets 1`] = `
"<root>
  <styles>
    <include src=\\"file://{resources}/styles/test.css\\"/>
  </styles>
  <scripts>
    <include src=\\"file://{resources}/scripts/test.js\\"/>
  </scripts>
  <snippets>
    <snippet name=\\"Ability\\">
      <Panel class=\\"Ability\\">
        <DOTAAbilityImage id=\\"AbilityImage\\"/>
      </Panel>
    </snippet>
    <snippet name=\\"AbilityLarge\\">
      <Panel class=\\"AbilityLarge\\">
        <DOTAAbilityImage id=\\"AbilityImage\\"/>
      </Panel>
    </snippet>
    <snippet name=\\"Item\\">
      <Panel class=\\"Item\\">
        <DOTAItemImage id=\\"ItemImage\\"/>
      </Panel>
    </snippet>
  </snippets>
  <Panel class=\\"test_root\\">
  </Panel>
</root>"
`;

exports[`Test xml render test.xml 1`] = `
"<root>
    <styles>
        <include src=\\"file://{resources}/styles/test.css\\"/>
    </styles>
    <scripts>
        <include src=\\"file://{resources}/scripts/test.js\\"/>
    </scripts>
    <snippets>
    </snippets>
    <Panel class=\\"test_root\\">
        <Panel id=\\"CustomElementA\\" class=\\"Demo DemoTest\\">
            <Panel id=\\"Header\\">
                <Label text=\\"Header of CustomElementA\\"/>
                <Button id=\\"CloseButton\\"/>
            </Panel>
            <Panel id=\\"Body\\">
                <Label text=\\"Body of CustomElementA\\"/>
                <Panel id=\\"CustomButton\\" onactivate=\\"OnCancel()\\">
                    <Image src=\\"file://{images}/custom_game/icons/cancel.png\\"/>
                    <Label text=\\"#DOTA_Cancel\\"/>
                </Panel>
            </Panel>
            <Panel id=\\"Footer\\">
                <Panel>
                    <Label text=\\"Footer of CustomElementA\\"/>
                </Panel>
            </Panel>
        </Panel>
        <Panel id=\\"CustomButton\\" onactivate=\\"OnClick()\\">
            <Image src=\\"file://{images}/custom_game/icons/plus.png\\"/>
            <Label text=\\"Button Text\\"/>
        </Panel>
        <Panel id=\\"PlusButton\\" onactivate=\\"OnClick()\\" class=\\"PlusButton\\">
            <Image src=\\"file://{images}/custom_game/icons/plus.png\\"/>
            <Label text=\\"Button Text\\"/>
        </Panel>
        <Panel id=\\"StyleButton\\" style=\\"background-color:#000;margin-top:5px;\\">
        </Panel>
    </Panel>
</root>"
`;

exports[`Test xml render test.xml no options 1`] = `
"<root>
  <styles>
    <include src=\\"file://{resources}/styles/test.css\\"/>
  </styles>
  <scripts>
    <include src=\\"file://{resources}/scripts/test.js\\"/>
  </scripts>
  <snippets>
  </snippets>
  <Panel class=\\"test_root\\">
    <Panel id=\\"CustomElementA\\" class=\\"Demo DemoTest\\">
      <Panel id=\\"Header\\">
        <Label text=\\"Header of CustomElementA\\"/>
        <Button id=\\"CloseButton\\"/>
      </Panel>
      <Panel id=\\"Body\\">
        <Label text=\\"Body of CustomElementA\\"/>
        <Panel id=\\"CustomButton\\" onactivate=\\"OnCancel()\\">
          <Image src=\\"file://{images}/custom_game/icons/cancel.png\\"/>
          <Label text=\\"#DOTA_Cancel\\"/>
        </Panel>
      </Panel>
      <Panel id=\\"Footer\\">
        <Panel>
          <Label text=\\"Footer of CustomElementA\\"/>
        </Panel>
      </Panel>
    </Panel>
    <Panel id=\\"CustomButton\\" onactivate=\\"OnClick()\\">
      <Image src=\\"file://{images}/custom_game/icons/plus.png\\"/>
      <Label text=\\"Button Text\\"/>
    </Panel>
    <Panel id=\\"PlusButton\\" onactivate=\\"OnClick()\\" class=\\"PlusButton\\">
      <Image src=\\"file://{images}/custom_game/icons/plus.png\\"/>
      <Label text=\\"Button Text\\"/>
    </Panel>
    <Panel id=\\"StyleButton\\" style=\\"background-color:#000;margin-top:5px;\\">
    </Panel>
  </Panel>
</root>"
`;

exports[`Test xml render test.xml templateRoots option 1`] = `
"<root>
  <styles>
    <include src=\\"file://{resources}/styles/test.css\\"/>
  </styles>
  <scripts>
    <include src=\\"file://{resources}/scripts/test.js\\"/>
  </scripts>
  <snippets>
  </snippets>
  <Panel class=\\"test_root\\">
    <Panel id=\\"CustomElementA\\" class=\\"Demo DemoTest\\">
      <Panel id=\\"Header\\">
        <Label text=\\"Header of CustomElementA\\"/>
        <Button id=\\"CloseButton\\"/>
      </Panel>
      <Panel id=\\"Body\\">
        <Label text=\\"Body of CustomElementA\\"/>
        <Panel id=\\"CustomButton\\" onactivate=\\"OnCancel()\\">
          <Image src=\\"file://{images}/custom_game/icons/cancel.png\\"/>
          <Label text=\\"#DOTA_Cancel\\"/>
        </Panel>
      </Panel>
      <Panel id=\\"Footer\\">
        <Panel>
          <Label text=\\"Footer of CustomElementA\\"/>
        </Panel>
      </Panel>
    </Panel>
    <Panel id=\\"CustomButton\\" onactivate=\\"OnClick()\\">
      <Image src=\\"file://{images}/custom_game/icons/plus.png\\"/>
      <Label text=\\"Button Text\\"/>
    </Panel>
    <Panel id=\\"PlusButton\\" onactivate=\\"OnClick()\\" class=\\"PlusButton\\">
      <Image src=\\"file://{images}/custom_game/icons/plus.png\\"/>
      <Label text=\\"Button Text\\"/>
    </Panel>
    <Panel id=\\"StyleButton\\" style=\\"background-color:#000;margin-top:5px;\\">
    </Panel>
  </Panel>
</root>"
`;
